src/crypto/tls/handshake_client_tls13.go

```diff
@@ -347,21 +350,40 @@ func (hs *clientHandshakeStateTLS13) processHelloRetryRequest() error {
 			c.sendAlert(alertIllegalParameter)
 			return errors.New("tls: server selected unsupported group")
 		}
-		if hs.ecdheParams.CurveID() == curveID {
+		if clientKeySharePrivateCurveID(hs.keySharePrivate) == curveID {
 			c.sendAlert(alertIllegalParameter)
 			return errors.New("tls: server sent an unnecessary HelloRetryRequest key_share")
 		}
 		if _, ok := curveForCurveID(curveID); curveID != X25519 && !ok {
```

Comment:
This check probably needs to be removed as it is duplicated for the `ecdhe` case below.

Due to this check, the `if scheme := curveIdToCirclScheme(curveID); scheme != nil {` check will not be reached either, so this code never seemed to be hit. Do we have a gap in our tests that need to be addressed? I only saw some ECH test failures, not other ones:
```
--- FAIL: TestECHHandshake (0.05s)
    --- FAIL: TestECHHandshake/11 (0.00s)
        ech_test.go:746: hrr / accepted
        ech_test.go:751: client aborts; want success
        ech_test.go:755: server aborts; want success
        ech_test.go:761: client err: remote error: tls: illegal parameter
        ech_test.go:767: server err: tls: client sent invalid key share in second ClientHello
```

Reply:
Good catch. This check was removed, but the bug persists, which suggests it's elsewhere.